tests/controllers/test_uaa_controller.py

```diff
@@ -282,3 +285,34 @@ def test_reset_user_password_by_id_password_change_success(self, mock_request):
             output = uaa_controller.change_user_password_by_id(user_id, user_password)
             self.assertEqual(output.status_code, 200)
             self.assertEqual(output.json(), password_change_success_json)
+
+    @requests_mock.mock()
+    def test_successful_processing_of_permissions_expiry_with_a_iso8601_date(self, mock_request):
+        mock_request.post(url_uaa_token, json={"access_token": self.access_token}, status_code=201)
+        mock_request.get(url_uaa_user_by_id, json=uaa_user_by_id_json, status_code=200)
+        with self.app.test_request_context():
+            session["permissions"] = {
+                "groups": test_groups,
+                "expiry": datetime.isoformat(datetime.now() + timedelta(minutes=-5)),
+            }
+            self.assertTrue(uaa_controller.user_has_permission("oauth.approvals", user_id))
+
+    @requests_mock.mock()
+    def test_successful_processing_of_permissions_expiry_with_a_datetime(self, mock_request):
+        mock_request.post(url_uaa_token, json={"access_token": self.access_token}, status_code=201)
+        mock_request.get(url_uaa_user_by_id, json=uaa_user_by_id_json, status_code=200)
+        with self.app.test_request_context():
+            session["permissions"] = {"groups": test_groups, "expiry": str(datetime.now() + timedelta(minutes=5))}
+            self.assertTrue(uaa_controller.user_has_permission("oauth.approvals", user_id))
+
+    @requests_mock.mock()
+    def test_unsuccessful_processing_of_permissions_expiry_with_an_incorrect_datetime_format(self, mock_request):
+        mock_request.post(url_uaa_token, json={"access_token": self.access_token}, status_code=201)
```

Comment:
You don't have to patch and assert a call here, however your mocks mock_request.post(url_uaa_token, json={"access_token": self.access_token}, status_code=201) etc are directly related to the refreshing function, which it won't reach because it falls over before then. (i.e they can be removed)

Reply:
This has been updated